cdc/sink/manager.go

```diff
@@ -0,0 +1,242 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sink
+
+import (
+	"context"
+	"math"
+	"sort"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/ticdc/cdc/model"
+)
+
+const (
+	// TODO: buffer chan size, the accumulated data is determined by
+	// the count of sorted data and unmounted data. In current benchmark a single
+	// processor can reach 50k-100k QPS, and accumulated data is around
+	// 200k-400k in most cases. We need a better chan cache mechanism.
+	defaultBufferChanSize = 1280000
+)
+
+// Manager manages table sinks, maintains the relationship between table sinks and backendSink
+type Manager struct {
+	backendSink  Sink
+	checkpointTs model.Ts
+	tableSinks   map[model.TableID]*tableSink
+	tableSinksMu sync.Mutex
+}
+
+// NewManager creates a new Sink manager
+func NewManager(ctx context.Context, backendSink Sink, errCh chan error, checkpointTs model.Ts) *Manager {
+	return &Manager{
+		backendSink:  newBufferSink(ctx, backendSink, errCh, checkpointTs),
+		checkpointTs: checkpointTs,
+		tableSinks:   make(map[model.TableID]*tableSink),
+	}
+}
+
+// CreateTableSink creates a table sink
+func (m *Manager) CreateTableSink(tableID model.TableID, checkpointTs model.Ts) Sink {
+	if _, exist := m.tableSinks[tableID]; exist {
+		log.Panic("the table sink already exists", zap.Uint64("tableID", uint64(tableID)))
+	}
+	sink := &tableSink{
+		tableID:   tableID,
+		manager:   m,
+		buffer:    make([]*model.RowChangedEvent, 0, 128),
+		emittedTs: checkpointTs,
+	}
+	m.tableSinksMu.Lock()
+	defer m.tableSinksMu.Unlock()
+	m.tableSinks[tableID] = sink
+	return sink
+}
+
+// Close closes the Sink manager and backend Sink
+func (m *Manager) Close() error {
+	return m.backendSink.Close()
+}
+
+func (m *Manager) getMinEmittedTs() model.Ts {
+	m.tableSinksMu.Lock()
+	defer m.tableSinksMu.Unlock()
+	if len(m.tableSinks) == 0 {
+		return m.getCheckpointTs()
+	}
+	minTs := model.Ts(math.MaxUint64)
+	for _, tableSink := range m.tableSinks {
+		emittedTs := tableSink.getEmittedTs()
+		if minTs > emittedTs {
+			minTs = emittedTs
+		}
+	}
+	return minTs
+}
+
+func (m *Manager) flushBackendSink(ctx context.Context) (model.Ts, error) {
+	minEmittedTs := m.getMinEmittedTs()
+	checkpointTs, err := m.backendSink.FlushRowChangedEvents(ctx, minEmittedTs)
+	if err != nil {
+		return m.getCheckpointTs(), errors.Trace(err)
+	}
+	atomic.StoreUint64(&m.checkpointTs, checkpointTs)
+	return checkpointTs, nil
+}
+
+func (m *Manager) destroyTableSink(tableID model.TableID) {
+	m.tableSinksMu.Lock()
+	defer m.tableSinksMu.Unlock()
+	delete(m.tableSinks, tableID)
+}
+
+func (m *Manager) getCheckpointTs() uint64 {
+	return atomic.LoadUint64(&m.checkpointTs)
+}
+
+type tableSink struct {
+	tableID   model.TableID
+	manager   *Manager
+	buffer    []*model.RowChangedEvent
+	emittedTs model.Ts
```

Comment:
Could you add some comments? Before update `emittedTs`, does it acquire receiving an ack from downstream?